hikari/impl/stateful_guild_chunker.py

```diff
@@ -23,65 +23,298 @@
 
 from __future__ import annotations
 
-__all__: typing.Final[typing.List[str]] = ["StatefulGuildChunkerImpl"]
+__all__: typing.Final[typing.List[str]] = ["StatefulGuildChunkerImpl", "ChunkStream"]
 
 import asyncio
-import logging
+import base64
+import copy
+import random
+import time
 import typing
 
+import attr
+
 from hikari import intents as intents_
+from hikari import snowflakes
+from hikari import undefined
 from hikari.api import chunker
-from hikari.impl import rate_limits
+from hikari.events import shard_events
+from hikari.utilities import attr_extensions
+from hikari.utilities import date
+from hikari.utilities import event_stream
+from hikari.utilities import mapping
 
 if typing.TYPE_CHECKING:
+    import datetime
+
     from hikari import guilds
     from hikari import traits
+    from hikari import users as users_
+    from hikari.api import shard as gateway_shard
+    from hikari.events import base_events  # noqa F401 - Unused (False positive)
+
+
+EXPIRY_TIME: typing.Final[int] = 5000
+"""How long a chunk event should wait until it's considered expired in miliseconds."""
+
+
+def _random_nonce() -> str:
+    head = time.perf_counter_ns().to_bytes(8, "big")
+    tail = random.getrandbits(92).to_bytes(12, "big")
+    return base64.b64encode(head + tail).decode("ascii")
+
+
+class ChunkStream(event_stream.EventStream[shard_events.MemberChunkEvent]):
+    """A specialised event stream used for triggering and streaming chunk events.
+
+    See Also
+    --------
+    Event Stream: `hikari.utilities.event_stream.EventStream`
+    """
+
+    __slots__: typing.Sequence[str] = (
+        "_guild_id",
+        "_include_presences",
+        "_limit",
+        "_missing_chunks",
+        "_nonce",
+        "_query",
+        "_users",
+    )
+
+    def __init__(
+        self,
+        app: traits.BotAware,
+        guild_id: snowflakes.Snowflake,
+        *,
+        timeout: typing.Union[int, float, None],
+        limit: typing.Optional[int] = None,
+        include_presences: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+        query_limit: int = 0,
+        query: str = "",
+        users: undefined.UndefinedOr[typing.Sequence[snowflakes.SnowflakeishOr[users_.User]]] = undefined.UNDEFINED,
+    ) -> None:
+        super().__init__(app=app, event_type=shard_events.MemberChunkEvent, limit=limit, timeout=timeout)
+        self._guild_id = guild_id
+        self._include_presences = include_presences
+        self._limit = query_limit
+        self._missing_chunks: typing.Optional[typing.MutableSequence[int]] = None
+        self._nonce = date.uuid()
```

Comment:
didnt you allow a custom nonce somewhere else?

Reply:
Nonces generated by the main StatefulGuildChunker implementation will follow the format "{shard_id}.{actual_nonce}" while nonces generated by the Chunk Stream will just be a uuid, didn't allow custom nonces anywhere as the StatefulGuildChunker includes metadata that's useful in the nonce and the Chunk Stream is less open than the guild chunker